doc_src/prompt.rst

```diff
@@ -0,0 +1,189 @@
+- :doc:`string join <cmds/string-join>`, if we want to have one thing appear between all our components.
+
+For now, let's use ``string join ''`` to write all our components without anything inbetween::
+
+  function fish_prompt
+      string join '' -- $PWD '>'
+  end
+
+The ``--`` tells ``string`` that no options can come after it, in case we extend this with something that can start with a ``-``.
+
+Adding color
+------------
+
+This prompt is functional, but a bit boring. We could add some color.
+
+In general, telling a terminal to print text in color involves sending it an escape sequence. For example you could send it
+
+.. role:: red
+.. parsed-literal::
+    :class: highlight
+
+    ESCAPE [ 3 1 m
+
+where "escape" is the ascii escape character stored as 0x1b.
+
+We could do this in fish, by writing escape as ``\e`` or ``\x1b``::
+
+  echo \e\[31mfoo
+
+This will tell the terminal to print "foo" in red and will look like this:
+
+.. parsed-literal::
+    :class: highlight
+
+    :red:`foo`
+
+To make it blue, we would use ``\e\[34mfoo``. However, obviously it's a bit annoying to have to remember the numbers for these colors, and modern terminals can do quite a bit more - most can not only do 256 colors but even 24-bit RGB.
+
+So, fish offers the :doc:`set_color <cmds/set_color>` command instead, so you can do::
+
+  echo (set_color red)foo
+
+``set_color`` can also handle RGB colors like ``set_color 23b455``, and other formatting options including bold and italics.
+
+So, taking our previous prompt and adding some color::
+
+  function fish_prompt
+      string join '' -- (set_color green) $PWD (set_color normal) '>'
+  end
+
+A "normal" color tells the terminal to go back to its normal formatting options.
+
+Shortening $PWD
+---------------
+
+This is alright, but our $PWD can be a bit long, and we are typically only interested in the last few directories. We can shorten this with the :doc:`prompt_pwd <cmds/prompt_pwd>` helper that will give us a shortened working directory::
+
+  function fish_prompt
+      string join '' -- (set_color green) (prompt_pwd) (set_color normal) '>'
+  end
+
+``prompt_pwd`` takes options to control how much to shorten. For instance if we want to display the last two directories, we'd use ``prompt_pwd --full-length-dirs=2`` or ``prompt_pwd -D 2``::
+
+  function fish_prompt
+      string join '' -- (set_color green) (prompt_pwd -D 2) (set_color normal) '>'
+  end
+
+With a current directory of "/home/tutorial/Music/Lena Raine/Oneknowing", this would print
+
+.. role:: green
+.. parsed-literal::
+    :class: highlight
+
+    :green:`~/M/Lena Raine/Oneknowing`>_
+
+Status
+------
+
+One important bit [#]_ of information that every command returns is the :ref:`status <variables-status>`. This is either 0 if the command returned successfully, or a whole number from 1 to 255 if not, so it's best understood as an error code.
```

Comment:
Not sure about the footnote - it will definitely make some people laugh but confuse others! The formal name for the status is "exit status" and we've tried to stick to that pretty consistently. Also non-0 statuses are not always errors! I'd replace the second sentence with something like "This is a number between 0 and 255, and 0 generally indicates that the command completed successfully, with other numbers used to indicate various error conditions".